File: datasets/python_source-code/Multiple-C1/organic_4.py
"""
Graph BFS - Organic Submission #4
BFS over an adjacency matrix representation instead of an adjacency list.
"""

from collections import deque


def bfs_matrix(matrix, start_index):
    n = len(matrix)
    visited = [False] * n
    visited[start_index] = True
    queue = deque([start_index])
    result = []

    while queue:
        curr = queue.popleft()
        result.append(curr)

        for neighbor_index in range(n):
            if matrix[curr][neighbor_index] == 1 and not visited[neighbor_index]:
                visited[neighbor_index] = True
                queue.append(neighbor_index)

    return result


if __name__ == "__main__":
    # 0:A 1:B 2:C 3:D 4:E
    adj_matrix = [
        [0, 1, 1, 0, 0],
        [1, 0, 0, 1, 0],
        [1, 0, 0, 0, 1],
        [0, 1, 0, 0, 1],
        [0, 0, 1, 1, 0],
    ]
    labels = ["A", "B", "C", "D", "E"]
    order = bfs_matrix(adj_matrix, 0)
    print([labels[i] for i in order])

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_c1/organic_4.java
/*
 * Graph BFS - Organic Submission #4
 * BFS over an adjacency matrix representation.
 */

import java.util.*;

class Solution {
    public static List<Integer> bfsMatrix(int[][] matrix, int startIndex) {
        int n = matrix.length;
        boolean[] visited = new boolean[n];
        visited[startIndex] = true;
        Queue<Integer> queue = new ArrayDeque<>();
        queue.add(startIndex);
        List<Integer> result = new ArrayList<>();

        while (!queue.isEmpty()) {
            int curr = queue.poll();
            result.add(curr);

            for (int neighborIndex = 0; neighborIndex < n; neighborIndex++) {
                if (matrix[curr][neighborIndex] == 1 && !visited[neighborIndex]) {
                    visited[neighborIndex] = true;
                    queue.add(neighborIndex);
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[][] adjMatrix = {
            {0, 1, 1, 0, 0},
            {1, 0, 0, 1, 0},
            {1, 0, 0, 0, 1},
            {0, 1, 0, 0, 1},
            {0, 0, 1, 1, 0}
        };
        String[] labels = {"A", "B", "C", "D", "E"};
        List<Integer> order = bfsMatrix(adjMatrix, 0);
        List<String> labeledOrder = new ArrayList<>();
        for (int idx : order) {
            labeledOrder.add(labels[idx]);
        }
        System.out.println(labeledOrder);
    }
}
